mypyc/genops.py

```diff
@@ -1126,23 +1213,153 @@ def generate_attr_defaults(self, cdef: ClassDef) -> None:
         self.functions.append(ir)
         cls.methods[ir.name] = ir
 
+    def load_non_ext_class(self, ir: ClassIR, line: int) -> Value:
+        assert self.non_ext_dict is not None
+        assert self.non_ext_anns is not None
+        assert self.non_ext_bases is not None
+
+        cls_name = self.load_static_unicode(ir.name)
+
+        # Add __annotations__ to the class dict.
+        self.primitive_op(dict_set_item_op,
+                          [self.non_ext_dict, self.load_static_unicode('__annotations__'),
+                           self.non_ext_anns], -1)
+
+        # We add a __doc__ attribute so if the non-extension class is decorated with the
+        # dataclass decorator, dataclass will not try to look for __text_signature__.
+        # https://github.com/python/cpython/blob/3.7/Lib/dataclasses.py#L957
+        filler_doc_str = 'filler docstring for classes decorated with dataclass'
+        self.primitive_op(dict_set_item_op,
+                         [self.non_ext_dict, self.load_static_unicode('__doc__'),
+                          self.load_static_unicode(filler_doc_str)], line)
+
+        metaclass = self.primitive_op(type_object_op, [], line)
+        metaclass = self.primitive_op(py_calc_meta_op, [metaclass, self.non_ext_bases], line)
+
+        class_type_obj = self.py_call(metaclass, [cls_name, self.non_ext_bases, self.non_ext_dict],
+                                      line)
+        return class_type_obj
+
+    def load_decorated_class(self, cdef: ClassDef, type_obj: Value) -> Value:
+        """
+        Given a decorated ClassDef and a register containing a non-extension representation of the
+        ClassDef created via the type constructor, applies the corresponding decorator functions
+        on that decorated ClassDef and returns a register containing the decorated ClassDef.
+        """
+        decorators = cdef.decorators
+        dec_class = type_obj
+        for d in reversed(decorators):
+            decorator = d.accept(self)
+            assert isinstance(decorator, Value)
+            dec_class = self.py_call(decorator, [dec_class], dec_class.line)
+        return dec_class
+
+    def populate_non_ext_bases(self, cdef: ClassDef) -> Value:
+        # Populate the base-class tuple passed to the metaclass constructor
+        # for non-extension classes.
```

Comment:
_Every_ multiline comment at the start of a function wants to be a docstring :-)